resources/bundles/org.eclipse.core.resources/src/org/eclipse/core/internal/resources/CharsetDeltaJob.java

```diff
@@ -76,6 +76,7 @@ interface ICharsetListenerFilter {
 	public CharsetDeltaJob(Workspace workspace) {
 		super(Messages.resources_charsetBroadcasting);
 		this.workspace = workspace;
+		setRule(workspace.getRoot()); // make sure workspace.prepareOperation() does not block
```

Comment:
I'm not sure if this is a good idea, just because I cannot foresee which code might be now affected by this job being waiting for a lock or other way around.

A more safe fix would be revert the part of the original change that introduced `join()`.

Reply:
While i am not 100% sure too effectivly the Jobs run() did wait on the workspace lock anyway. If cancel+join is removed we would be back to the old exception or - depending how far the job is  - the warning that a job is still running ... the plugin should cancel it before shutdown. The fix just did follow the guidelines :-)

Reply:
My point is: I've never heard about issues on shutdown before, so this seem to be a corner case and from what I understand just of "cosmetic" kind. Not nice, but doesn't harm anyone.

To fix a minor issue, we added the join() and run in deadlocks. 
This is not minor issue anymore.

To fix deadlocks, we add more locks due the rule. 
This will most likely lead to more issues at runtime, as the job will run at different times due the rule and other jobs will run also at different times because waiting on this rule taken.

I'm not sure if all these changes with all the possible implications are worth to fix a minor issue logged at shutdown.

Reply:
It always was only a problem on junits, the deadlock only occur on junits too. The benefit to fix minor logging issues in tests is that the logfiles get rid of unrelated issues so that its easier to find the real issues.
https://github.com/eclipse-platform/eclipse.platform/blob/c5d730d4ae3fc16418df8165b7f46cb36daddf82/runtime/bundles/org.eclipse.core.jobs/src/org/eclipse/core/internal/jobs/JobManager.java#L724
is very explict that the implementation was wrong. Keeping a wrong implementation can also mislead others to copy&paste wrong patterns.

Reply:
That is not nice, and `cancel()` may help here. 
But `join()` to get rid of the message? It is simply not adequate solution, the side effects are worse as the benefit.

Reply:
Depending on how long one assumes the job should run a *timed join* might be better as it at least would exit somehow in case of deadlock situation.

Reply:
Even though there is no precedent (all oher stop/shutdowns like egit/jdt use cancel+join) a timeout seems to be a valid compromise. I updated the PR.

Reply:
But with timeout we don't need the rule anymore. Could we please remove it?

Reply:
Without the rule it would needlessly wait the timeout time whenever it deadlocked before. The rule prevents that the timeout is ever reached (if the computer does not lag) and is used in similar Jobs (for example AutoBuildJob, ApiAnalysisMarkersJob, FlushJob ... ) as well. Please note that the CharsetDeltaJob is internal only and does nothing more then create Markers and update the int ResourceInfo.charsetAndContentId. I don't see any dangerous things or third party code that is called from there.

Reply:
This looks good in general. However, in `AutoBuildJob` and `FlushJob` the rule is set in the Job constructor but also inside the `run()` method the same rule is used for the workspace operation. Should `CharsetDeltaJob` also do this?

Reply:
That looks unnecessary superfluous to me. I don't know why it is like that.